### ISSUE TITLE
arXiv '22 | Using DeepSpeed and Megatron to Train Megatron-Turing NLG 530B, A Large-Scale Generative Language Model

### ISSUE THREAD

**Issue by vycezhong** [vycezhong/read-papers#261]
https://arxiv.org/pdf/2201.11990.pdf

3D parallelism巅峰了.

**Comment by vycezhong** [comment]
和blog内容差不多. https://developer.nvidia.com/blog/using-deepspeed-and-megatron-to-train-megatron-turing-nlg-530b-the-worlds-largest-and-most-powerful-generative-language-model/

还是那一套训练方法. 没有太多创新. 当然，这一整套流程做下来，是非常非常厉害的. 

训练的是NLG，所以采用transformer decoder. 

Takeaway: 
- 560个DGX A100 nodes, 总共**4480 A100 GPUs**（显存80GB），每个node有8个NVIDIA Mellanox 200Gbps HDR Infiniband HCAs，还有两个额外的HCAs用于存储. datacenter采用three-level (leaf, spine, core) fat-tree topology总共850个switches. 集群采用all-NVME shared parallel filesystem.
- 8-way tensor-slicing within a node, 35-way pipeline parallelism，那么16-way data parallelism. (t,p,d) = (8, 35, 16)
- mixed precision使用bfloat16
- seq_len = 2048. total_batch_size = 1920. 他们在训练过程中会慢慢增加batch size. 前1B tokens进行linear warmup. lr = 5e-5，使用cosine decay. Adam optimizer (\beta_1 = 0.9, \beta_2 = 0.95). grad clip norm = 1, weight decay = 0.1. 这些参数应该是参考了GPT-3 #56 . 
- 使用了270B tokens进行训练. 


看看结果.

首先是LAMBADA数据集，这是一个对话段落数据集，会给一个context和一句需要补全的话. 它的特点是必须要看context才能做出补全. 

这有几个例子. 
![image](https://user-images.githubusercontent.com/25879526/154194913-1ee7b174-7286-4c7c-9180-45fc9bb98604.png)

MT-NLG结果当然是SOTA了. 不过我感觉提升相比GPT-3不是特别大，考虑到MT-NLG模型大小都是GPT-3三倍了. 可能训练真的太难了，很多参数不好调. 
![image](https://user-images.githubusercontent.com/25879526/154194931-e24d476a-a29d-4180-b17a-571ecc2af3ef.png)
